pkg/expression/contextopt/sessionvars.go

```diff
@@ -0,0 +1,57 @@
+// Copyright 2024 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package contextopt
+
+import (
+	"github.com/pingcap/tidb/pkg/expression/context"
+	"github.com/pingcap/tidb/pkg/sessionctx/variable"
+	"github.com/pingcap/tidb/pkg/util/intest"
+)
+
+var _ context.OptionalEvalPropProvider = &SessionVarsPropProvider{}
+var _ RequireOptionalEvalProps = SessionVarsPropReader{}
+
+// SessionVarsPropProvider is a provider to get the session variables
+type SessionVarsPropProvider struct {
+	variable.SessionVarsProvider
+}
+
+// NewSessionVarsProvider returns a new SessionVarsPropProvider
+func NewSessionVarsProvider(provider variable.SessionVarsProvider) *SessionVarsPropProvider {
+	intest.AssertNotNil(provider)
+	return &SessionVarsPropProvider{provider}
+}
+
+// Desc implements the OptionalEvalPropProvider interface.
+func (p *SessionVarsPropProvider) Desc() *context.OptionalEvalPropDesc {
+	return context.OptPropSessionVars.Desc()
+}
+
+// SessionVarsPropReader is used by expression to read property context.OptPropSessionVars
+type SessionVarsPropReader struct{}
+
+// RequiredOptionalEvalProps implements the RequireOptionalEvalProps interface.
+func (SessionVarsPropReader) RequiredOptionalEvalProps() context.OptionalEvalPropKeySet {
+	return context.OptPropSessionVars.AsPropKeySet()
+}
+
+// GetSessionVars returns the session vars from the context
+func (SessionVarsPropReader) GetSessionVars(ctx context.EvalContext) (*variable.SessionVars, error) {
+	p, err := getPropProvider[*SessionVarsPropProvider](ctx, context.OptPropSessionVars)
```

Comment:
The key itself must exist, what is the point of adding this judgment? The main thing is that an extra error judgment feels that the code is not very elegant.

Reply:
If this is an impossible situation, is it better to just panic instead of return an error?

Specifically, when will a return error occur?

Reply:
In the future, we may have multiple implementations for `EvalContext` and not all of them are required to contain all properties. The expression should fail when a property is not provided but required by an expression.

Usually, the error occurs when a coder constructs a `EvalContext` but missing to provide some required properties or using some unsupported expression in a specified scene. For example, only some simple expressions are allowed in a partition expression, but if someone uses function `GetVar`, it will fail.

Though in the above partition expression case, we should do some prior checks but we should also have a check in `GetSessionVars` to make the code strong. IMO, panic is not a good option because it will make tidb (at leat a session) crash.

Reply:
Is it to be changed in the future, and panic is handled at present?